CredentialProvider.Microsoft/CredentialProviders/Vsts/IAuthUtil.cs

```diff
@@ -73,17 +81,25 @@ public async Task<Uri> GetAadAuthorityUriAsync(Uri uri, CancellationToken cancel
             return new Uri($"{aadBase}/common");
         }
 
-        public async Task<bool> IsVstsUriAsync(Uri uri)
+        public async Task<IFeedUriSource> GetFeedUriSource(Uri uri)
         {
-            if (!IsValidScheme(uri))
+            // Ping the url to see from headers whether it's an Azure Artifacts feed or external
+            var responseHeaders = await GetResponseHeadersAsync(uri, cancellationToken: default);
+
+            // Hosted only allows https
+            if (IsHttpsScheme(uri) && responseHeaders.Contains(VssResourceTenant) && responseHeaders.Contains(VssAuthorizationEndpoint))
             {
-                // We are not talking to a https endpoint so it cannot be a VSTS endpoint
-                return false;
+                return IFeedUriSource.Hosted;
             }
 
-            var responseHeaders = await GetResponseHeadersAsync(uri, cancellationToken: default);
+            // Uri returns 401 (TFS service error) if it's an Azure DevOps Server instance.
+            if (responseHeaders.Contains(TfsServiceError))
```

Comment:
This is the header I'd like to have a discussion about

Reply:
I would use a different header. I don't think we want to specifically identify an error case here. `X-VSS-E2EID` should be on all responses